### PR TITLE
[#888] Added interface CreditBasedSender with methods to check credits and to set a drain handler

### PR DESCRIPTION
**Pull request by kaniyan**
Added interface `CreditBasedSender` with methods to check credits and to set a drain handler. With reference to the issue #888, this PR also provides possibility for the command client to retrieve the _available credits_ and also to set a _drain handler_.

### CODE REVIEW

**Comment by kaniyan** [comment]
It sounds even better. I will go for it :+1:

File: client/src/main/java/org/eclipse/hono/client/CreditBasedSender.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     * with more credit from the peer. For subsequent notifications to be received, a new handler must be registered.
+     * <p>
+     * Client code can use this method to register a handler after it has checked this sender's capacity to send
+     * messages using {@link #getCredit()}, e.g.
+     * 
+     * <pre>
+     * MessageSender sender;
+     * ...
+     *
+     * sender.send(msg);
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/CreditBasedSender.java]: this doesn't make much sense here, does it?

File: client/src/main/java/org/eclipse/hono/client/MessageSender.java [hunk 2]
```diff
@@ -27,70 +27,30 @@
...
-     * 
-     * @return The number of messages.
-     */
-    int getCredit();
-
     /**
      * Checks if this sender can send or buffer (and send later) a message.
      * <p>
-     * Note that the value returned is valid during execution of the current vert.x handler only.
+     * Note that the value returned is valid durin execution of the current vert.x handler only.
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/MessageSender.java]: during

File: tests/src/test/java/org/eclipse/hono/tests/http/HttpTestBase.java [hunk 3]
```diff
@@ -896,6 +898,109 @@ public void testUploadMessagesWithTtdThatReplyWithCommand(final TestContext ctx)
...
 
+
+    /**
+     * Verifies that the queueDrainHandler is invoked on available of credit and a command is sent.
+     *
+     * @param ctx The test context.
+     * @throws InterruptedException if the test fails.
+     */
+    @Test
+    public void testQueueDrainHandler(final TestContext ctx) throws InterruptedException {
```
Comment (sophokles73) [inline tests/src/test/java/org/eclipse/hono/tests/http/HttpTestBase.java]: this test looks very big in comparison to the very little functionality that we actually want to verify. Given that all of the relevant functionality has already been verified in the unit tests, I believe that we can safely remove this test.

Comment (kaniyan) [reply tests/src/test/java/org/eclipse/hono/tests/http/HttpTestBase.java]: Done.

File: client/src/main/java/org/eclipse/hono/client/MessageSender.java (hunk 2 above)
Comment (kaniyan) [reply client/src/main/java/org/eclipse/hono/client/MessageSender.java]: Thanks for noticing it. It has been corrected.

File: client/src/main/java/org/eclipse/hono/client/CreditBasedSender.java (hunk 1 above)
Comment (kaniyan) [reply client/src/main/java/org/eclipse/hono/client/CreditBasedSender.java]: I removed this part

File: client/src/main/java/org/eclipse/hono/client/MessageSender.java [hunk 4]
```diff
@@ -27,70 +27,30 @@
  * A client for publishing messages to Hono.
  *
  */
-public interface MessageSender {
+public interface MessageSender extends CreditBasedSender{
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/MessageSender.java]: missing whitespace before opening braces

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 5]
```diff
@@ -321,6 +322,33 @@ public final void setRequestTimeout(final long timoutMillis) {
         }
     }
 
+    @Override
+    public int getCredit() {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: should be final

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 6]
```diff
@@ -321,6 +322,33 @@ public final void setRequestTimeout(final long timoutMillis) {
...
+        if (sender == null) {
+            return 0;
+        } else {
+            return sender.getCredit();
+        }
+    }
+
+    @Override
+    public void sendQueueDrainHandler(final Handler<Void> handler) {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: should be final

File: tests/src/test/java/org/eclipse/hono/tests/http/HttpTestBase.java [hunk 7]
```diff
@@ -896,6 +895,7 @@ public void testUploadMessagesWithTtdThatReplyWithCommand(final TestContext ctx)
                 });
     }
 
+
```
Comment (sophokles73) [inline tests/src/test/java/org/eclipse/hono/tests/http/HttpTestBase.java]: there is no real change to this file, is there? If so, I think it should be removed from this commit.

File: client/src/main/java/org/eclipse/hono/client/MessageSender.java (hunk 4 above)
Comment (kaniyan) [reply client/src/main/java/org/eclipse/hono/client/MessageSender.java]: I already run the checkstyle plugin. It didn't report any error. I will change this.

File: tests/src/test/java/org/eclipse/hono/tests/http/HttpTestBase.java [hunk 8]
```diff
@@ -163,7 +163,6 @@ public static void init(final TestContext ctx) {
      */
     @Before
     public void setUp() {
-
```
Comment (sophokles73) [inline tests/src/test/java/org/eclipse/hono/tests/http/HttpTestBase.java]: can you please remove this change from the commit?

Comment (kaniyan) [reply tests/src/test/java/org/eclipse/hono/tests/http/HttpTestBase.java]: This file is no more included in commits.

File: tests/src/test/java/org/eclipse/hono/tests/IntegrationTestSupport.java [hunk 9]
```diff
@@ -343,6 +341,23 @@ public String getRandomDeviceId(final String tenantId) {
...
 
+    /**
+     * Returns a command client for a device.
+     *
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The identifier of the device.
+     * @param requestTimeout The number of milliseconds to wait for a response from the device.
+     * @return A future that may contain a command client.
+     */
+    public Future<CommandClient> getCommandClient(final String tenantId, final String deviceId,
```
Comment (sophokles73) [inline tests/src/test/java/org/eclipse/hono/tests/IntegrationTestSupport.java]: this change also seems to be unrelated to this PR, or isn't it? If it is, please remove it.
